### ISSUE TITLE
haproxy solaris compile include/haproxy/queue-t.h:43:8: error: redefinition of ‘struct queue’  gcc version 13.2.0

### ISSUE THREAD

**Issue by geerttouquet** [haproxy/haproxy#2449]
### Detailed Description of the Problem

 gmake TARGET=solaris USE_STATIC_PCRE=1 USE_PCRE=1 USE_OPENSSL=1 USE_ZLIB=1 USE_CRYPT_H=1 USE_LIBCRYPT=1 CC=/usr/bin/gcc
  CC      src/ev_poll.o
  CC      src/ev_evports.o
  CC      src/ssl_sock.o
In file included from include/haproxy/server-t.h:37,
                 from include/haproxy/lb_map-t.h:26,
                 from include/haproxy/backend-t.h:30,
                 from include/haproxy/proxy-t.h:34,
                 from include/haproxy/obj_type.h:32,
                 from include/haproxy/stream.h:29,
                 from include/haproxy/channel.h:30,
                 from include/haproxy/applet.h:29,
                 from src/ssl_sock.c:47:
include/haproxy/queue-t.h:43:8: error: redefinition of ‘struct queue’
   43 | struct queue {
      |        ^~~~~
compilation terminated due to -Wfatal-errors.

### Expected Behavior

that you can compile the code

### Steps to Reproduce the Behavior

just try to compile

### Do you have any idea what may have caused this?

?

### What is your configuration?

```haproxy
gcc version 13.2.0 (GCC)
GNU Make 4.3
```

### Output of `haproxy -vv`

```plain
can't compile
```

**Comment by git001** [comment]
Please can you share which version of HAProxy do you want to compile?  
Please can you share which OS version you have?

**Comment by geerttouquet** [comment]
ybase@voyager ~/custom/install_ddl/PROD_S9> uname -a
SunOS voyager 5.11 11.4.63.157.1 sun4v sparc sun4v logical-domain

**Comment by Darlelet** [comment]
probably duplicate of #1794

**Comment by wtarreau** [comment]
So indeed from what I'm seeing all over the net, Solaris apparently decided that the generic word "queue" that appears in english dictionaries, belongs to them and that no program may ever use it :-(   Unfortunately my Sun died a few years ago so I can't test anymore. But from what I've seen in some reports, it would be useful to test with combinations of `DEFINE=-D_XPG6`, `DEFINE=-D__EXTENSIONS__=1`, or `DEFINE=-DXOPEN_SOURCE=600`, and share the output so that we see if that's getting us closer or farther from a solution.

And please share your haproxy version, we still have no idea what version you're trying to compile!

**Comment by chipitsine** [comment]
I'm trying current haproxy master using the following "[qemu in linux](https://github.com/vmactions)" actions:

https://github.com/chipitsine/haproxy/actions/runs/7923059299/job/21631988366

(I'll play with mentioned defines in a week)

**Comment by wtarreau** [comment]
I didn't know there were solaris VMs on GH actions, that's good to know!

**Comment by chipitsine** [comment]
hopefully, we can enable monthly builds of solaris/netbsd/openbsd, it should not burn much electricity on running vm in vm

**Comment by geerttouquet** [comment]
I will try to compile vesion 2.9.4

**Comment by geerttouquet** [comment]
geert@voyager /usr/local/src/haproxy/haproxy-2.9.4> gmake TARGET=solaris CPU=ultrasparc USE_STATIC_PCRE=1  USE_OPENSSL=1 CC=/usr/bin/gcc DEFINE=-D_XPG6
  CC      src/slz.o
  CC      src/ev_poll.o
  CC      src/ev_evports.o
  CC      src/ssl_sock.o
In file included from include/haproxy/server-t.h:37,
                 from include/haproxy/lb_map-t.h:26,
                 from include/haproxy/backend-t.h:30,
                 from include/haproxy/proxy-t.h:34,
                 from include/haproxy/obj_type.h:32,
                 from include/haproxy/stream.h:29,
                 from include/haproxy/channel.h:30,
                 from include/haproxy/applet.h:29,
                 from src/ssl_sock.c:47:
include/haproxy/queue-t.h:43:8: error: redefinition of 'struct queue'
   43 | struct queue {
      |        ^~~~~
compilation terminated due to -Wfatal-errors.
gmake: *** [Makefile:1020: src/ssl_sock.o] Error 1

**Comment by geerttouquet** [comment]
geert@voyager /usr/local/src/haproxy/haproxy-2.9.4> gmake TARGET=solaris CPU=ultrasparc USE_STATIC_PCRE=1  USE_OPENSSL=1 CC=/usr/bin/gcc DEFINE=-D__EXTENSIONS__=1
  CC      src/slz.o
  CC      src/ev_poll.o
  CC      src/ev_evports.o
  CC      src/ssl_sock.o
In file included from include/haproxy/server-t.h:37,
                 from include/haproxy/lb_map-t.h:26,
                 from include/haproxy/backend-t.h:30,
                 from include/haproxy/proxy-t.h:34,
                 from include/haproxy/obj_type.h:32,
                 from include/haproxy/stream.h:29,
                 from include/haproxy/channel.h:30,
                 from include/haproxy/applet.h:29,
                 from src/ssl_sock.c:47:
include/haproxy/queue-t.h:43:8: error: redefinition of 'struct queue'
   43 | struct queue {
      |        ^~~~~
compilation terminated due to -Wfatal-errors.
gmake: *** [Makefile:1020: src/ssl_sock.o] Error 1
geert@voyager /usr/local/src/haproxy/haproxy-2.9.4> gmake TARGET=solaris CPU=ultrasparc USE_STATIC_PCRE=1  USE_OPENSSL=1 CC=/usr/bin/gcc DEFINE=-DXOPEN_SOURCE=600
  CC      src/slz.o
  CC      src/ev_poll.o
  CC      src/ev_evports.o
  CC      src/ssl_sock.o
In file included from include/haproxy/server-t.h:37,
                 from include/haproxy/lb_map-t.h:26,
                 from include/haproxy/backend-t.h:30,
                 from include/haproxy/proxy-t.h:34,
                 from include/haproxy/obj_type.h:32,
                 from include/haproxy/stream.h:29,
                 from include/haproxy/channel.h:30,
                 from include/haproxy/applet.h:29,
                 from src/ssl_sock.c:47:
include/haproxy/queue-t.h:43:8: error: redefinition of 'struct queue'
   43 | struct queue {
      |        ^~~~~
compilation terminated due to -Wfatal-errors.

**Comment by geerttouquet** [comment]
is there an option to change the name queue to another name ?

**Comment by wtarreau** [comment]
Thanks for the tests. I guess it won't be easy :-(

No there's no option to change the internal struct name, that's a breaking change which would cause difficulties for backporting fixes, and would even cause trouble to external components or patches that currently know this structure. Other projects have faced the same issue, some apparently just giving up on Solaris and others seemingly finding their way through various such options. But their approach of defining this struct is absurd, it's almost as bad as if they redefined main().

**Comment by geerttouquet** [comment]
thanks for the suggestion.  other suggestion will be welcome

**Comment by geerttouquet** [comment]
We are running solaris on sparc servers M8

**Comment by geerttouquet** [comment]
If I change 'struct queue' to   'struct queueTemp' , then I can compile it with following option
  gmake TARGET=solaris CPU=ultrasparc   USE_OPENSSL=1 CC=/usr/bin/gcc

**Comment by wtarreau** [comment]
OK that's a good news already, it means that it's the only one causing trouble.

I had a look and it looks like queue-t is only included from a few files after the system includes, so maybe it will work to add this just before "struct queue" in queue-t.h:

  #if defined(sun)
  #define queue _queue
  #endif

That would be dirty, but if it works I'd be fine with it. If it does not work (other conflicts, etc), then you can try to add this at the end of "include/haproxy/compat.h", just before the last "#endif". This file is also made to maintain compatibility between various OSes and it would be nice there.

**Comment by geerttouquet** [comment]
This fix works fine, perhaps we can this fix include in the official release ?
#if defined(sun)
#define queue _queue
#endif

**Comment by wtarreau** [comment]
OK thank you. Probably that the best thing to do is to commit this one and backport it, then plan to rename it for next versions and get rid of the ifdef.

**Comment by geerttouquet** [comment]
Can I do something for this, or will you do this ?

**Comment by wtarreau** [comment]
If you feel at ease with authoring a patch that mentions this issue, feel free to attach it here and I'll gladly merge it. Please have a look at CONTRIBUTING to see how to format the subject and explain the problem so that in the future if we want to drop it we can re-evaluate the consequences. If you don't feel at ease with this, no problem, we'll do it as time permits :-)

**Comment by mx1up** [comment]
I gave it a try, hope I did it right 😅 
```
From 9b892aa85aa21987685301808d1c6f411d51c48e Mon Sep 17 00:00:00 2001
From: matthias sweertvaegher <178714+mx1up@users.noreply.github.com>
Date: Fri, 23 Feb 2024 17:05:01 +0100
Subject: [PATCH] BUILD: solaris: fix redefinition of queue struct
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Compilation on solaris fails because the name 'queue' is reserved:
> include/haproxy/queue-t.h:43:8: error: redefinition of ‘struct queue’

see issue https://github.com/haproxy/haproxy/issues/2449

This patch redefines 'queue' as '_queue' which fixes compilation for now.

Future plan: rename 'queue' in code base so define can be removed again.
---
 include/haproxy/queue-t.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/include/haproxy/queue-t.h b/include/haproxy/queue-t.h
index 8f6a1ec8d..93ab00a1b 100644
--- a/include/haproxy/queue-t.h
+++ b/include/haproxy/queue-t.h
@@ -40,6 +40,12 @@ struct pendconn {
        __decl_thread(HA_SPINLOCK_T del_lock);  /* use before removal, always under queue's lock */
 };

+
+/* redefine on solaris, as it is now a reserved name */
+#if defined(sun)
+#define queue _queue
+#endif
+
 struct queue {
        struct eb_root head;                    /* queued pendconnds */
        struct proxy  *px;                      /* the proxy we're waiting for, never NULL in queue */
--
2.39.3 (Apple Git-145)

```

**Comment by wtarreau** [comment]
Thanks. Please try to place it at the end of include/haproxy/compat.h as I suggested above, it would be much cleaner since that's where we place the compatibility stuff. It should work the same but I prefer to be certain that it won't cause build failures anywhere else as a side effect.

In your commit message you can indicate what versions this should be backported to. Since there were very few reports about this, I'm going to limit the risks and probably do 2.9 and 2.8 only. But I'm not strict on this, if others need it, we can discuss.

Also, please attach the file next time so that tabs are preserved. If you prefer to do a PR, you can and I'll pick the patch there. It's just a bit more work on both sides.

Last, I know that we have a few github e-mail addresses in a few commit messages, but it's better to use a real one so that the author can be contacted later, e.g. if the patch causes trouble and another approach needs to be tested.

And you were right to mention the future plan in the patch. I think that after this we'll rename the field before 3.0 is released so as to get rid of this for the long term.

Thanks!

**Comment by mx1up** [comment]
Yes, I understand, thanks for the feedback. I created a PR but @geerttouquet will first need to validate the patch before we can proceed. He will probably be able to test monday.

As for a contact: I think @geerttouquet , being the sys admin, will be the better person, but I'll let him decide. Our email addresses are firstname.lastname@nexuzhealth.com

Should I mention 3.0 in the future plan?

thanks

**Comment by wtarreau** [comment]
OK perfect! There's no emergency. Your current commit message is fine, no need to go deeper into the details. Let's wait for Geert on monday then. Thanks!

**Comment by geerttouquet** [comment]
I have try the following
git clone https://github.com/mx1up/haproxy.git
git checkout 20240223-fix-solaris-compilation

I have try to compile it and got following error now (I don't got in version 2.9.5
In file included from include/haproxy/xxhash.h:43,
                 from src/server.c:49:
include/import/xxhash.h:3390:42: warning: "__cplusplus" is not defined, evaluates to 0 [-Wundef]
 3390 | #if ((defined(sun) || defined(__sun)) && __cplusplus) /* Solaris includes __STDC_VERSION__ with C++. Tested with GCC 5.5 */
      |                                          ^~~~~~~~~~~
In file included from /usr/include/sys/socket.h:32,
                 from include/haproxy/compat.h:30,
                 from include/haproxy/api-t.h:35,
                 from include/import/ebtree-t.h:24,
                 from include/import/ebtree.h:250,
                 from include/import/ebmbtree.h:25,
                 from src/server.c:19:
src/server.c: In function ‘snr_resolution_cb’:
src/server.c:4441:40: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘.’ token
 4441 |                 struct server_inetaddr s_addr;
      |                                        ^~~~~~
compilation terminated due to -Wfatal-errors.
gmake: *** [Makefile:1006: src/server.o] Error 1

**Comment by wtarreau** [comment]
Very strange, there's no "." on this line. I'm starting to worry that "s_addr" is also #defined on solaris :-( I had memories of an OS where almost everything used to compile fine but nowadays it seems their OS is crippled with generic names that conflict with those that applications choose.

This s_addr is indeed new in 3.0 from what I can check, so it seems we'll need to rename it. Could you please try to rename "s_addr" to "srv_addr" in this function, at lines 4441..4445, 4455..4459, 4541..4545 ? If so that's something we can quickly change.

Regarding the c++ warning, it should be addressed like this, but I'll check if it's already fixed in the latest xxhash as I don't like to modify imported files:
```
--- a/include/import/xxhash.h
+++ b/include/import/xxhash.h
@@ -3387,7 +3387,7 @@ XXH_PUBLIC_API XXH64_hash_t XXH64_hashFromCanonical(XXH_NOESCAPE const XXH64_can
 
 /* ===   Compiler specifics   === */
 
-#if ((defined(sun) || defined(__sun)) && __cplusplus) /* Solaris includes __STDC_VERSION__ with C++. Tested with GCC 5.5 */
+#if ((defined(sun) || defined(__sun)) && defined(__cplusplus) && __cplusplus) /* Solaris includes __STDC_VERSION__ with C++. Tested with GCC 5.5 */
 #  define XXH_RESTRICT   /* disable */
 #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L   /* >= C99 */
 #  define XXH_RESTRICT   restrict
```

**Comment by geerttouquet** [comment]
I have change the code you have suggested and it compiles on solaris :-) 
rename "s_addr" to "srv_addr" in server.c and it works fine

**Comment by mx1up** [comment]
I have added the changes on the branch (see new PR) and rebased it on master.
I updated the commit message to reflect the additional changes.

I await your confirmation whether the `__cplusplus` define warning needs to be addressed in this PR as well or not.

**Comment by mx1up** [comment]
@chipitsine any progress / luck with the solaris CI build setup?

**Comment by geerttouquet** [comment]
I have compile the branch with succes on solaris

**Comment by chipitsine** [comment]
> @chipitsine any progress / luck with the solaris CI build setup?

https://github.com/chipitsine/haproxy/actions/runs/7923059299/job/21631988366

**Comment by wtarreau** [comment]
OK thanks guys, I'll have a look.

**Comment by wtarreau** [comment]
With Matthias' patch merged, we can now close.

**Comment by wtarreau** [comment]
Hmm, no, let's mark it for 2.9 before closing.

**Comment by ThrawnCA** [comment]
This still affects 2.6, which is meant to be in LTS.

**Comment by wtarreau** [comment]
Indeed, 2.6 also has this struct queue. Could you please check that backporting commit aeb5b08df3 ("BUILD: solaris: fix compilation errors") from 2.8 fixes the problem for you ?

**Comment by ThrawnCA** [comment]
> Could you please check that backporting commit [aeb5b08](https://github.com/haproxy/haproxy/commit/aeb5b08df381d4797780c878f7e7774ecaa13124) ("BUILD: solaris: fix compilation errors") from 2.8 fixes the problem for you ?

That does fix the build, yes. I had to dig around a bit to figure out exactly where to put it, since "CLOCK_MONOTONIC_COARSE" doesn't exist in 2.6, but eventually I decided it could still go immediately before `#endif /* _HAPROXY_COMPAT_H */`
